db/attachment.go

```diff
@@ -218,32 +236,40 @@ type AttachmentCallback func(name string, digest string, knownData []byte, meta
 // its data. The callback is told whether the attachment body is known to the database, according
 // to its digest. If the attachment isn't known, the callback can return data for it, which will
 // be added to the metadata as a "data" property.
-func (db *Database) ForEachStubAttachment(body Body, minRevpos int, callback AttachmentCallback) error {
+func (db *Database) ForEachStubAttachment(body Body, minRevpos int, docID string, callback AttachmentCallback) error {
 	atts := GetBodyAttachments(body)
 	if atts == nil && body[BodyAttachments] != nil {
-		return base.HTTPErrorf(400, "Invalid _attachments")
+		return base.HTTPErrorf(http.StatusBadRequest, "Invalid _attachments")
 	}
 	for name, value := range atts {
 		meta, ok := value.(map[string]interface{})
 		if !ok {
-			return base.HTTPErrorf(400, "Invalid attachment")
+			return base.HTTPErrorf(http.StatusBadRequest, "Invalid attachment")
 		}
 		if meta["data"] == nil {
 			if revpos, ok := base.ToInt64(meta["revpos"]); revpos < int64(minRevpos) || !ok {
 				continue
 			}
 			digest, ok := meta["digest"].(string)
 			if !ok {
-				return base.HTTPErrorf(400, "Invalid attachment")
+				return base.HTTPErrorf(http.StatusBadRequest, "Invalid attachment")
 			}
-			data, err := db.GetAttachment(AttachmentKey(digest))
-			if err != nil && !base.IsDocNotFoundError(err) {
-				return err
+			version, ok := GetAttachmentVersion(meta)
```

Comment:
As this PR stands, the incoming `body` is never going to have the attachment version set.  We've discussed a subsequent enhancement to retrieve the current revision from the bucket in order to be able to properly compute revpos, but if we're at this point and there's an attachment on the incoming document with revpos < minRevpos, then that's an attachment Sync Gateway doesn't know about, and should be stored in the new format.

Reply:
This logic has been refactored. It assumes the attachment is always AttVersion2 while checking whether it has already been uploaded. CBG-1590 needs to be implemented to determine whether incoming attachment is AttVersion1 or AttVersion2.